cpp/oneapi/dal/table/detail/csr.hpp

```diff
@@ -0,0 +1,255 @@
+
+    using base = is_table_impl<T>;
+
+    static constexpr bool value = base::template has_method_get_column_count_v<T> &&
+                                  base::template has_method_get_row_count_v<T> &&
+                                  base::template has_method_get_metadata_v<T> &&
+                                  base::template has_method_get_data_layout_v<T> &&
+                                  has_method_get_data_v<T> && has_method_get_column_indices_v<T> &&
+                                  has_method_get_row_indices_v<T>;
+};
+
+template <typename T>
+inline constexpr bool is_csr_table_impl_v = is_csr_table_impl<T>::value;
+
+} // namespace v1
+
+using v1::is_csr_table_impl;
+using v1::is_csr_table_impl_v;
+
+namespace v1 {
+
+class ONEDAL_EXPORT sparse_table {
+    friend detail::pimpl_accessor;
+    using pimpl = detail::pimpl<detail::sparse_table_impl_iface>;
+
+public:
+    /// An empty table constructor: creates the table instance with zero number of rows and columns.
+    /// Implementation is set to the special "empty" object that returns all the property values
+    /// set to default (see Properties section).
+    sparse_table();
+
+    /// Creates a new table instance that shares the implementation with another one.
+    sparse_table(const sparse_table&) = default;
+
+    /// Replaces the implementation by another one.
+    sparse_table& operator=(const sparse_table&) = default;
+
+    /// Indicates whether a table contains non-zero number of rows and columns.
+    bool has_data() const noexcept;
+
+    /// The number of columns in the table.
+    /// @remark default = 0
+    std::int64_t get_column_count() const;
+
+    /// The number of rows in the table.
+    /// @remark default = 0
+    std::int64_t get_row_count() const;
+
+    /// The metadata object that holds additional information
+    /// about the data within the table.
+    /// @remark default = table_metadata()
+    const table_metadata& get_metadata() const;
+
+    /// The runtime id of the table type.
+    /// Each table sub-type has its unique ``kind``.
+    /// An empty table (see the default constructor) has a unique ``kind`` value as well.
+    /// @remark default = empty_table_kind
+    std::int64_t get_kind() const;
+
+    /// The layout of the data within the table
+    /// @remark default = data_layout::csr
+    data_layout get_data_layout() const;
+
+protected:
+    sparse_table(pimpl::element_type* impl);
+    void sparse_init_impl(pimpl::element_type* impl);
+
+private:
+    pimpl impl_;
+};
+
+class ONEDAL_EXPORT csr_table : public sparse_table {
+    friend detail::pimpl_accessor;
+    using pimpl = detail::pimpl<dal::backend::csr_table_impl>;
+
+public:
+    /// Creates a new ``csr_table`` instance with zero number of rows and columns.
+    /// The :expr:`kind` is set to``csr_table::kind()``.
+    /// All the properties should be set to default values (see the Properties section).
+    csr_table();
```

Comment:
General question: for homogen table, we have row-major and column-major support inside. It's worth considering be consistent here - will it bring benefits having `csx_table` that can store `csr` or `csc` data?